lib/process_wrapper/system_posix.cc

```diff
@@ -0,0 +1,157 @@
+
+// posix headers
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include <iostream>
+#include <vector>
+
+namespace process_wrapper {
+
+namespace {
+
+class StdoutPipe {
+ public:
+  static constexpr size_t kReadEndDesc = 0;
+  static constexpr size_t kWriteEndDesc = 1;
+
+  ~StdoutPipe() {
+    CloseReadEnd();
+    CloseWriteEnd();
+  }
+
+  int CreateEnds() {
+    if (pipe(stdout_pipe_desc_) != 0) {
+      std::cerr << "process wrapper error: failed to open the stdout pipes."
+                << std::endl;
+      return false;
+    }
+    return true;
+  }
+  void DupWriteEnd() {
+    dup2(stdout_pipe_desc_[kWriteEndDesc], STDOUT_FILENO);
+    CloseReadEnd();
+    CloseWriteEnd();
+  }
+
+  void CloseReadEnd() { Close(kReadEndDesc); }
+  void CloseWriteEnd() { Close(kWriteEndDesc); }
+
+  int ReadEndDesc() const { return stdout_pipe_desc_[kReadEndDesc]; }
+  int WriteEndDesc() const { return stdout_pipe_desc_[kWriteEndDesc]; }
+
+  bool WriteToFile(const System::StrType& stdout_file) {
+    CloseWriteEnd();
+
+    constexpr size_t kBufferSize = 4096;
+    char buffer[kBufferSize];
+    int stdout_file_desc =
+        open(stdout_file.c_str(), O_WRONLY | O_CREAT | O_TRUNC,
+             S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+    if (stdout_file_desc == -1) {
+      std::cerr << "process wrapper error: failed to open redirection file."
+                << std::endl;
+      return false;
+    }
+    while (1) {
+      ssize_t read_bytes = read(ReadEndDesc(), buffer, kBufferSize);
+      if (read_bytes < 0) {
+        std::cerr << "process wrapper error: failed child process stdout."
+                  << std::endl;
+        return false;
+      } else if (read_bytes == 0) {
+        break;
+      }
+      ssize_t written_bytes = write(stdout_file_desc, buffer, read_bytes);
+      if (written_bytes < 0 || written_bytes != read_bytes) {
+        std::cerr << "process wrapper error: failed to write to stdout file."
+                  << std::endl;
+        return false;
+      }
+    }
+
+    CloseReadEnd();
+    close(stdout_file_desc);
+    return true;
+  }
+
+ private:
+  void Close(size_t idx) {
+    if (stdout_pipe_desc_[idx] > 0) {
+      close(stdout_pipe_desc_[idx]);
+    }
+    stdout_pipe_desc_[idx] = -1;
+  }
+  int stdout_pipe_desc_[2] = {-1};
+};
+
+}  // namespace
+
+System::StrType System::GetWorkingDirectory() {
+  const size_t kMaxBufferLength = 4096;
+  char cwd[kMaxBufferLength];
+  if (getcwd(cwd, sizeof(cwd)) == NULL) {
+    return System::StrType{};
+  }
+  return System::StrType{cwd};
+}
+
+int System::Exec(const System::StrType& executable,
+                 const System::Arguments& arguments,
+                 const System::EnvironmentBlock& environment_block,
+                 const StrType& stdout_file) {
+  StdoutPipe stdout_pipe;
+  if (!stdout_file.empty() && !stdout_pipe.CreateEnds()) {
+    return -1;
+  }
+
+  pid_t child_pid = fork();
```

Comment:
Why do we need to fork at all? From what I see in the description of this wrapper, there is no need to do anything upon process termination, so the wrapper can set things up and then give control to the wrapper process, keeping the process tree cleaner (and being much faster e.g. on macOS).

Reply:
The touch file feature, needs getting back the control.